`New Feature` Termed Roles

dacebt opened the pull request:
Adds DecentAutonomousAdmin Contract
- Used for trigging next start and flush any sablier funds to previous elected wearer

Adds DecentHats_0_2_0 Contract
- DecentAutoAdmin clone is deployed and added as wearer of Admin Hat
- Updated to create Termed Roles when prompted
- This involves making the deploying a instance of Hats Election Module and setting the eligibility module

Closes #103 
Closes #104 
Closes #105

dacebt committed:
add getHatEligibilityModule to IHats interface

dacebt committed:
Add HatsElectionEligibility interface

dacebt committed:
wip DecentAutonomousAdminHat contract

dacebt committed:
create AA Hat

dacebt committed:
remove unused param

dacebt committed:
update interfaces with needed methods

dacebt committed:
add tests for Wearer requirement

dacebt committed:
add deployment file

dacebt committed:
drop hat off name

dacebt committed:
add Autonomous Admin as wearer of admin hat
- break out create method into seperate function

dacebt committed:
remove only

dacebt committed:
fix deploy name

dacebt committed:
pass eligibility address as param

dacebt committed:
update with proxy pattern

dacebt committed:
update with new props to deploy proxy

dacebt committed:
rename and update tests to support proxy pattern

dacebt committed:
remove only

dacebt committed:
update DecentHats tests to support changes to contract

dacebt committed:
update deploy with constructor arg

dacebt committed:
use existing salt

dacebt committed:
remove only

dacebt committed:
don't add new variables

dacebt committed:
add interface for HatModuleFactory

dacebt committed:
update to add electionModuleAddress to termed roles

dacebt committed:
update test with new params

adamgall commented on contracts/DecentHats_0_1_0.sol:
We're going to need to create another version of this contract now (0_2_0), because 

1. this existing version has already been deployed on networks _and_ has been used to create real roles on real daos
2. this contract's address is part of the salt used to determine role's smartAccount addresses.

So this means we can't just update this and deploy a new one, because then we'll be determining the incorrect smartAccount address for any existing Roles on DAOs that used this current version to create their Roles.

dacebt replied:
thats a good point. going to have to figure out a good way to enable versioning of our contracts. Noting for later: What does this transaction to 'update' look like.

dacebt committed:
rename param

dacebt committed:
revert use of moduleProxyfactory to deploy DecentAutoAdmin

dacebt committed:
rename termEndTs -> termEndDateTs

dacebt committed:
Update tests to support changes

dacebt committed:
revert DecentHats_0_1_0 contract

dacebt committed:
remove versioning from constructor

dacebt committed:
remove unneeded deploy file

dacebt committed:
add DecentHats_0_2_0 contract and deployment script

dacebt committed:
reorder structs for gas efficiency

dacebt committed:
revert 0_1_0 test and add 0_2_0 tests

dacebt committed:
update test

dacebt committed:
fix deployment script

dacebt committed:
use ModuleProxyFactory to deploy AutoAdmin

dacebt committed:
update tests to support changes

adamgall commented on contracts/DecentHats_0_1_0.sol:
```diff
@@ -247,4 +247,4 @@ contract DecentHats_0_1_0 {
         params.hatsProtocol.transferHat(topHatId, address(this), msg.sender);
     }
-}
+}
```
Please make sure this change is reverted.

adamgall commented on contracts/interfaces/hats/IHats.sol:
```diff
@@ -39,4 +39,15 @@ interface IHats {
     ) external returns (bool success);
 
     function transferHat(uint256 _hatId, address _from, address _to) external;
+
+    function getHatEligibilityModule(
+        uint256 _hatId
+    ) external view returns (address eligibility);
+
+    function isWearerOfHat(
+        address _user,
+        uint256 _hatId
+    ) external view returns (bool isWearer);
+
+    function changeHatEligibility(uint256 _hatId, address _newEligibility) external;
```
Please see the following commits and discussions from other PRs:

- Relevant discussion, but for Sablier not Hats: https://github.com/decentdao/decent-contracts/pull/100#discussion_r1792349087
- Relevant commit which implements what I think should happen in this branch, too: https://github.com/decentdao/decent-contracts/pull/108/commits/e1fd7d75a9ba03aa9f47c4fef03a8cf3f359eace

Please let me know if it's not clear why we need to do this. The TL;DR is that if we _don't_, then you'll notice that running the deployment scripts will result in a new instance of `DecentHats_0_1_0` being deployed (because the bytecode has changed), and we need to avoid that.

dacebt replied:
AH you know I watched for that and got fooled by the caching and hadn't looked at the deployment diff yet ha. cool beans I'll take a look at what you've done and update accordingly

adamgall replied:
I actually implemented it on @DarksightKellar's branch... let me cherry pick a couple of commits over

dacebt replied:
even better 🙏

adamgall replied:
@Da-Colon holy shit it ended up being 17 commits LMAO. Had to revert the changes to both the Hats interfaces _and_ the Sablier interfaces. Luckily, the two other open branches implement the "fixes" for both Hats and Sablier, so more of less that part was already done just had to cherry pick them over.

Then, a lot of other commits happened due to cleanup. But that's only because I made changes very atomically. Recommend you read through them in order to see what I did.

adamgall committed:
Add PRBMath to package, for Sablier interfaces

adamgall committed:
Added minimum amount of "full" Sablier interfaces needed for DecentSablierStreamManagement, to also be future-proof

adamgall committed:
Downtype "sablier" in SablierStreamParams to an address, and rename it to sablierV2LockupLinear

adamgall committed:
Fix MockSablierV2LockupLinear contract to actually implement full ISablierV2Lockup interface

adamgall committed:
Use new "full" Sablier interfaces in DecenHats_0_2_0

adamgall committed:
Type the sablier address as the "full" ISablierV2Lockup type in DecentAutonomousAdmin

adamgall committed:
Remove changes from original ISablierV2LockupLinear interface, so no bytecode changes to DecentHats_0_1_0 occur

adamgall committed:
Named imports in DecentAutonomousAdmin

adamgall committed:
Copy in the FULL Hats interfaces

adamgall committed:
Rename IHatModuleFactory to IHatsModuleFactory, place it in Hats directory, make sure the interface and mock implementation are complete

adamgall committed:
Move IHatsElectionsElegibility into "full" directory

adamgall committed:
If you don't need to use a variable in a function just delete the name of it in the input params

adamgall committed:
Finish implementing full Hats interface in MockHatsAdmin

adamgall committed:
Use "full" IHats interface in DecentAutonomousAdmin and DecentHats_0_2_0

adamgall committed:
Revert changes to original IHats interface

adamgall committed:
Revert change to DecentHats_0_1_0 contract -- this actually does cause a difference in compiled bytecode!

adamgall committed:
Remove unused import from DecentHats_0_2_0

adamgall committed:
Support solidity version 0.8.28 in hardhat config

adamgall committed:
Merge branch 'develop' into add-contract/DecentAutonomousAdminHat

adamgall committed:
Update Sablier contracts to their actual solidity pragma

adamgall committed:
Use solidity version 0.8.28 for new contracts

dacebt commented on contracts/DecentHats_0_2_0.sol:
```diff
@@ -11,7 +11,6 @@ import {LockupLinear, Broker} from "./interfaces/sablier/full/types/DataTypes.so
 import {DecentAutonomousAdmin} from "./DecentAutonomousAdmin.sol";
 import {IHatsModuleFactory} from "./interfaces/hats/full/IHatsModuleFactory.sol";
 import {IHatsElectionEligibility} from "./interfaces/hats/full/IHatsElectionEligibility.sol";
-import {Clones} from "@openzeppelin/contracts/proxy/Clones.sol";
```
ty

dacebt commented on contracts/DecentAutonomousAdmin.sol:
```diff
@@ -1,5 +1,6 @@
 // SPDX-License-Identifier: MIT
-pragma solidity =0.8.19;
+pragma solidity 0.8.28;
```
👍

dacebt commented on contracts/DecentHats_0_1_0.sol:
```diff
@@ -247,4 +247,4 @@ contract DecentHats_0_1_0 {
     }
-}
\ No newline at end of file
+}
```
👍

dacebt reviewed (commented):
coolio changes makes sense. Question though. Should we just change the 'deploy' script to prevent any new deployments of that contract.

Another question. Should we probably think about a different pattern for updates for V2 and beyond?

dacebt committed:
renaming and move salt to constant

dacebt committed:
update tests to read SALT from contract

dacebt committed:
Merge branch 'develop' of github.com:decent-dao/fractal-contracts into add-contract/DecentAutonomousAdminHat

dacebt committed:
reorder new deploy scripts

adamgall commented on contracts/DecentAutonomousAdmin.sol:
```diff
@@ -0,0 +1,130 @@
+
+contract DecentAutonomousAdmin {
+    string public constant NAME = "DecentAutonomousAdmin";
+    string public version_ = "0.1.0";
```
Should probably make this `constant`

adamgall commented on contracts/DecentAutonomousAdmin.sol:
```diff
@@ -0,0 +1,130 @@
+import {ISablierV2Lockup} from "./interfaces/sablier/full/ISablierV2Lockup.sol";
+
+contract DecentAutonomousAdmin {
+    string public constant NAME = "DecentAutonomousAdmin";
+    string public version_ = "0.1.0";
```
But actually...

For this contract, which is a "long standing" contract (aka not an ephemeral contract in the context of a proposal execution that we add/remove as a module to execute specific logic), I'd like to see ERC165 support built in. Will be good for having a programmatic way to determine contract functionality when this contract will inevitably be upgraded in the future, and we may have DAOs with different versions of this contract installed on their Admin role.

I think the ERC165 support should replace the (custom, non-idiomatic) `NAME` and `version_` variables.

dacebt replied:
Just saw these comments over the weekend. Let me get this done

adamgall commented on contracts/DecentHats.sol:
```diff
@@ -0,0 +1,324 @@
+import {ModuleProxyFactory} from "@gnosis.pm/zodiac/contracts/factory/ModuleProxyFactory.sol";
+
+contract DecentHats {
+    string public constant NAME = "DecentHats";
```
Same comments here. Instead of this NAME variable, I'd like to see ERC165 support built into this contract.

adamgall replied:
Wait nevermind we don't need ERC165 in here, this is one of those "ephemeral" contracts.

adamgall replied:
We also probably just don't even need this NAME string in the contract either. Like who cares, yanno? Idk why I added it in DecentHats_0_1_0.

adamgall replied:
fixed

adamgall commented on deploy/core/011_deploy_ModuleProxyFactory.ts:
I would prefer you undo the changes to this file.

adamgall commented on deploy/core/017_deploy_DecentHats_0_1_0.ts:
I'd prefer you follow the same pattern as script 011 (after addressing my comment)

adamgall commented on test/DecentHats_0_1_0.test.ts:
Undo this change please.

dacebt committed:
revert changes made to struct

dacebt committed:
set eligibility module and elect when new term role

dacebt committed:
fix tests

dacebt committed:
remove flushing of payments when triggering new term

DarksightKellar reviewed (commented):
Man, I started reviewing this some time ago and I remember I had some questions at some point, but now I'm back to it and I don't remember what. Since this is significant work, I'd love to have an all hands engineering review at some point so we can all, both 
 - come to the same page on what the feature+code is and should look like, and
 - contribute and pick apart in real time, where thoughts will bounce of each other and flow a lot more freely.

dacebt committed:
termed role payments are direct to nominated wearer

dacebt committed:
clean up
- seperate out logic for minting termed roles

dacebt committed:
make factoryFriendly

dacebt committed:
revert changes

dacebt committed:
revert changes

dacebt committed:
change for consistency

dacebt committed:
revert change

dacebt commented on contracts/DecentAutonomousAdmin.sol:
```diff
@@ -0,0 +1,61 @@
+    // //////////////////////////////////////////////////////////////
+    //                         initializer
+    // //////////////////////////////////////////////////////////////
+    function setUp(bytes memory initializeParams) public override initializer {}
```
Uses the same patterns as the modules deployed via `ModuleProxyFactory` which requires a setUp function with params, but in its current state we don't need it

dacebt committed:
add comments

dacebt commented on contracts/DecentAutonomousAdmin.sol:
```diff
@@ -0,0 +1,61 @@
+    // //////////////////////////////////////////////////////////////
+    //                         Public Functions
+    // //////////////////////////////////////////////////////////////
+    function triggerStartNextTerm(TriggerStartArgs calldata args) public {
```
Pretty straight forward. The Contract allows the Admin Hat, when worn to help trigger the next term. While the next term can be triggered by anyone. we need a few other things to happen when this occurs for UX purposes.

- After starting next term. The Original Hat should be burned via `checkHatWearerStatus`. if ineligibile the hat is burned. Then we just mint a new hat to the new wearer

dacebt committed:
small cleanup

dacebt commented on contracts/DecentHats.sol:
```diff
@@ -0,0 +1,395 @@
+
+        _updateKeyValuePairs(params.keyValuePairs, topHatId);
+
+        (uint256 adminHatId, ) = _createAdminHatAndAccount(
+            params.hatsProtocol,
+            params.registry,
+            params.moduleProxyFactory,
+            params.decentAutonomousAdminMasterCopy,
+            params.hatsAccountImplementation,
+            topHatAccount,
+            topHatId,
+            params.adminHat
+        );
```
This function was created from `_createHatAndAccountAndMintAndStreams` and narrowed down to just Admin hat creation.

dacebt commented on contracts/DecentHats.sol:
```diff
@@ -0,0 +1,393 @@
+        for (uint256 i = 0; i < params.hats.length; ) {
+            if (params.hats[i].isTermed) {
+                // Create election module and set as eligiblity
+                _createTermedHatAndAccountAndMintAndStreams(
+                    params.hatsProtocol,
+                    topHatAccount,
+                    _createElectionEligiblityModule(
+                        params.hatsModuleFactory,
+                        params.hatsElectionEligibilityImplementation,
+                        params.hatsProtocol.getNextId(adminHatId),
+                        topHatId,
+                        params.hats[i].termedParams[0]
+                    ),
+                    adminHatId,
+                    params.hats[i]
+                );
```
This is the other 'new' workflow to review

dacebt committed:
add ERC165 support

adamgall committed:
Merge branch 'develop' into add-contract/DecentAutonomousAdminHat

adamgall committed:
Lint and pretty

adamgall commented on contracts/interfaces/IDecentAutonomousAdmin.sol:
```diff
@@ -0,0 +1,21 @@
+import {ISablierV2Lockup} from "./sablier/full/ISablierV2Lockup.sol";
+
+interface IDecentAutonomousAdmin {
+    struct SablierStreamInfo {
+        uint256 streamId;
+        ISablierV2Lockup sablierV2Lockup;
+    }
```
I think this struct (and the `ISablierV2Lockup`) are unused and can be removed

dacebt replied:
🤦 oof yeah good call

adamgall commented on contracts/mocks/MockHatsAdmin.sol:
Nitpicky but can you rename this file to MockHatsAutoAdmin to match the contract name within

adamgall replied:
Also, what's the reason for this specific mock contract? It seems like it's supposed to mock the core Hats contract. Any reason why you didn't just add this code to the existing MockHats contract?

dacebt replied:
I'll take another look at the AutoAdmin test and Mock contracts it uses now with everything updated

adamgall replied:
fixed

adamgall reviewed (changes requested):
High-level first-draft review

adamgall committed:
Merge branch 'develop' into add-contract/DecentAutonomousAdminHat

adamgall commented on contracts/DecentHats.sol:
```diff
@@ -0,0 +1,393 @@
+                    _createElectionEligiblityModule(
+                        params.hatsModuleFactory,
+                        params.hatsElectionEligibilityImplementation,
+                        params.hatsProtocol.getNextId(adminHatId),
```
Has this been tested when there are multiple termed hats being created?

edit: actually I do think this would work as written, yeah. `getNextId(adminHatId)` is being called after each new hat is created, yeah.

adamgall replied:
but still, definitely need some tests written

dacebt replied:
got the tree test. Need to add a test for `createTermedRoleHat` external function

adamgall replied:
addressed

adamgall commented on test/DecentHats.test.ts:
Nowhere in this test are there any hats being created where `isTermed: true`. That seems like an oversight...

adamgall replied:
@Da-Colon I remember you saying yesterday that there was a reason why there are no tests testing `isTermed: true`. Can you explain that again?

adamgall reviewed (changes requested):
Please write some tests for the new DecentHats contract where termed hats are being created.

dacebt committed:
remove unused structand unsued comment area

dacebt committed:
remove unneeded conditionals

dacebt committed:
add requires to check term param length and nominees length

DarksightKellar committed:
Update natspec

adamgall commented on contracts/DecentHats.sol:
Make sure that the code from most recent `DecentHats_0_1_0` gets implemented into this contract.

https://github.com/decentdao/decent-contracts/blob/develop/contracts/DecentHats_0_1_0.sol#L198-L233

@DarksightKellar please provide @Da-Colon with the proper natspec comments, too

dacebt replied:
done + updated

dacebt committed:
Merge branch 'develop' of github.com:decent-dao/fractal-contracts into add-contract/DecentAutonomousAdminHat

dacebt committed:
update with public function to create new RoleHats

dacebt committed:
Add createTermRole external function
- update params into structs and calldata

dacebt committed:
switch createAndDeclareTree function to external

dacebt committed:
move Create Sablier Stream logic to internal function

dacebt committed:
fix sablier recipient for role hats

dacebt committed:
update tests to latest
- matches DecentHats test to original contract
- adds eslint ignore for ethers import for type-safety

dacebt committed:
update eligibility mock contract to more align with live contract

dacebt committed:
add `getNextId` method

dacebt committed:
add 'create tree with term roles' tests

dacebt commented on test/DecentHats.test.ts:
```diff
@@ -0,0 +1,826 @@
+import { SignerWithAddress } from '@nomicfoundation/hardhat-ethers/signers';
+import { expect } from 'chai';
+/* eslint-disable-next-line import/no-extraneous-dependencies */
+import { ethers } from 'ethers';
```
@adamgall @DarksightKellar So that 'can't find ethers namespace' errors was removing type safety. After quite a bit of trial and error of trying things. I found this is the simplest solution. `hardhat` or one of the `hardhat` libraries is adding `ethers` to our `node_modules` but it is not in our `package.json` but adding it there conflicts with the version used in the hard hat environment. 

So doing this allows for it to properly find it and doesn't prevent the other errors from showing up. 

I did not add this to other files from my own as I didn't want to fill up this PR but I can.

adamgall replied:
I dig it, thanks. I'll open up a standalone PR.

adamgall replied:
Well actually there are only two other test files that have this issue. Would you mind just applying this fix to those two other files in your current PR? thanks.

dacebt replied:
Ah good call thought it was in more places, but looks like its only effecting where we are trying to pull `types` out of the package

dacebt committed:
eslint ignore ethers import

DarksightKellar committed:
Put natspec update in the correct contract

DarksightKellar committed:
Cleanup natspec

DarksightKellar committed:
Add natspec clarifying stream recipient difference between termed and untermed roles

DarksightKellar commented on test/DecentAutonomousAdmin.test.ts:
```diff
@@ -0,0 +1,106 @@
+        sablierStreamInfo: [], // No Sablier stream info for this test
+      };
+
+      // Verify the hat is now worn by the current wearer
```
Incorrect copy-paste comment

DarksightKellar commented on test/DecentAutonomousAdmin.test.ts:
```diff
@@ -0,0 +1,106 @@
+      await adminHat.triggerStartNextTerm(args);
+
+      // Verify the hat is now worn by the nominated wearer
+      expect((await hatsProtocol.isWearerOfHat(nominatedWearer.address, userHatId)) === true);
```
Suggest adding another test to verify the previous wearer of the hat was "removed"

dacebt committed:
fix typo

dacebt committed:
expect the hat wearer to lose hat

dacebt committed:
add test for `createTermedRoleHat`

adamgall commented on contracts/DecentHats_0_1_0.sol:
Why are there changes in this contract @Da-Colon @DarksightKellar ??

This contract has already been deployed and it's not going to be deployed again (see the commented out deployment code in deployer file 017).

We should not be making changes to contract source code if there's no intention of deploying it. We should make sure that whatever source code is checked into source control always contains the most-recently-deployed contract on public chains.

DarksightKellar replied:
Reverted. The natspec update went onto the wrong contract

adamgall replied:
Thank you

adamgall commented on contracts/interfaces/IDecentAutonomousAdmin.sol:
```diff
@@ -0,0 +1,16 @@
+pragma solidity 0.8.28;
+
+import {IHats} from "./hats/full/IHats.sol";
+import {ISablierV2Lockup} from "./sablier/full/ISablierV2Lockup.sol";
```
Does this import need to be here? It's not being referenced in this interface file, so I would say this is probably not the right place for it.

adamgall replied:
fixed

DarksightKellar commented on test/DecentHats.test.ts:
```diff
@@ -0,0 +1,970 @@
+      });
+    });
+
+    describe('Creating a new hat on existing Tree', () => {
```
Should be "Creating a new Termed hat on existing tree"

adamgall replied:
fixed

DarksightKellar committed:
Revert misplaced natspec update

adamgall commented on contracts/mocks/MockHats.sol:
Considering that this Mock contract implements the not-"`full`" `IHats` interface, which is specifically _only_ used in an old version of `DecentHats_0_1_0`, I'm wondering why there needs to be additional functionality added to this Mock contract?

adamgall commented on contracts/mocks/MockHats.sol:
Why do we need additional functionality in this contract? This specific Mock implements the non-"`full`" `IHats` interface, which is only used in the old `DecentHats_0_1_0` contract. I do see that there are failing tests if I remove these new mocked functions, but that leads me to believe that the test file is relying on this `MockHats` instance when maybe it should be relying on the new `MockHatsAutoAdmin` instance.

adamgall replied:
fixed

adamgall commented on contracts/mocks/MockHatsAdmin.sol:
Please make sure the filename of this contract matches the actual contract name.

adamgall replied:
fixed

adamgall commented on contracts/mocks/MockHatsAdmin.sol:
Is there any reason why this Mock file / contract has the word "Admin" in it?

Maybe a more appropriate name would be something like "MockHatsFull", to refer to the fact that it implements the "full" IHats interface.

adamgall replied:
fixed

adamgall committed:
Remove unused import from interface

adamgall committed:
Consolidate Mock Hats into a single contract

adamgall committed:
Update test name

adamgall commented on contracts/DecentAutonomousAdmin.sol:
```diff
@@ -0,0 +1,51 @@
+    //                         Public Functions
+    // //////////////////////////////////////////////////////////////
+    function triggerStartNextTerm(TriggerStartArgs calldata args) public {
+        require(
+            args.hatsProtocol.isWearerOfHat(args.currentWearer, args.hatId),
+            "Not current wearer"
+        );
```
To stick with modern solidity, can we use actual `error`s here?

dacebt replied:
fixed

adamgall commented on contracts/DecentAutonomousAdmin.sol:
```diff
@@ -0,0 +1,51 @@
+            args.hatsProtocol.isWearerOfHat(args.currentWearer, args.hatId),
+            "Not current wearer"
+        );
+        address hatsEligibilityModuleAddress = args
+            .hatsProtocol
+            .getHatEligibilityModule(args.hatId);
```
To save some gas, let's not allocate memory for this variable, since it's only being used in one place. Just inline this into line 33.

dacebt replied:
fixed

adamgall commented on contracts/DecentHats.sol:
```diff
@@ -0,0 +1,485 @@
+import {IERC6551Registry} from "./interfaces/IERC6551Registry.sol";
+import {IHats} from "./interfaces/hats/full/IHats.sol";
+import {LockupLinear, Broker} from "./interfaces/sablier/full/types/DataTypes.sol";
+import {DecentAutonomousAdmin} from "./DecentAutonomousAdmin.sol";
```
This import seems unneeded.

adamgall replied:
fixed

adamgall commented on contracts/DecentHats.sol:
```diff
@@ -0,0 +1,485 @@
+        TermedParams[] termedParams;
+        uint32 maxSupply;
+        bool isMutable;
+        bool isTermed;
```
Do we need to have this boolean? Can we not infer it based on the existence of any items in the `termedParams` array?

adamgall replied:
fixed, thanks

adamgall commented on contracts/DecentHats.sol:
```diff
@@ -0,0 +1,485 @@
+        string details;
+        string imageURI;
+        SablierStreamParams[] sablierParams;
+        TermedParams[] termedParams;
```
Why is this an array?

adamgall replied:
addressed

adamgall commented on contracts/DecentHats.sol:
```diff
@@ -0,0 +1,485 @@
+        electionModuleAddress = hatsModuleFactory.createHatsModule(
+            hatsElectionEligibilityImplementation,
+            hatId,
+            abi.encode(topHatId, uint256(0)),
```
Does this 0 need to be cast? Also, can you please add some comments on these lines about what these "immutable" and "mutable" args represent, from the PoV of the Elections Eligibility contract?

adamgall commented on contracts/DecentHats.sol:
```diff
@@ -0,0 +1,485 @@
+        require(
+            hat.termedParams[0].nominatedWearers.length == 1,
+            "DecentHats: nominatedWearers length must be 1"
+        );
```
Can we use actual `error`s here? And everywhere in this contract.

adamgall replied:
fixed

adamgall commented on contracts/DecentHats.sol:
```diff
@@ -0,0 +1,485 @@
+            hat.termedParams[0].nominatedWearers
+        );
+
+        hatsProtocol.mintHat(hatId, hat.termedParams[0].nominatedWearers[0]);
```
Definitely open to pushback on this suggestion, but, what if we called `batchMintHat` here and passed in the entire `nominatedWearers` array? Leave it to the frontend to just make sure we're only passing in one of them. That might allow us to continue using this same contract if we do ever decide to open up "role members" to > 1 at a time.

adamgall replied:
Edit: nevermind, this would cause a discrepancy in how we're setting up streams then, cause we are hardcoding the [0] there. And it's beyond us to decide how streams should happen if there are multiple nominated wearers.

So now maybe I'm thinking, can we just turn `nominatedWearers` into a single `address`? We'd need to put that address into an array when calling eligibility.elect, but that's fine. I kind of like that better, making the contract's input params match our business logic.

adamgall replied:
addressed this

adamgall commented on contracts/DecentHats.sol:
```diff
@@ -0,0 +1,485 @@
+        IHatsElectionEligibility(eligibilityAddress).elect(
+            hat.termedParams[0].termEndDateTs,
+            hat.termedParams[0].nominatedWearers
+        );
```
Nitpick, flip the order of these two statements.

dacebt replied:
?

dacebt replied:
I guess it doesn't really matter, but it seem to make sense to create the hat then elect?

adamgall replied:
yeahh... my nitpick is just because the `hatId` variable isn't used in the `elect` call, but it's used in the line below that. My intention with the nitpick comment was just to keep relevant code close to each other.

> I guess it doesn't really matter, but it seem to make sense to create the hat then elect?

but the `elect`ion doesn't depend on the new `hatId`, is what i'm saying.

dacebt committed:
review feedback

adamgall commented on contracts/DecentHats.sol:
```diff
@@ -0,0 +1,481 @@
+        );
+    }
+
+    function _createHat(
+        IHats _hatsProtocol,
+        uint256 adminHatId,
+        Hat memory _hat,
+        address toggle,
+        address eligibility
+    ) internal returns (uint256) {
+        return
+            _hatsProtocol.createHat(
+                adminHatId,
+                _hat.details,
+                _hat.maxSupply,
+                eligibility,
+                toggle,
+                _hat.isMutable,
+                _hat.imageURI
+            );
+    }
```
Is it benefitting us to have this single external function call wrapped in an internal function? Why not just inline the `_hatsProtocol.crateHat()` call wherever it's used?

dacebt replied:
fixed.

adamgall commented on contracts/DecentHats.sol:
```diff
@@ -0,0 +1,481 @@
+            );
+    }
+
+    function _createAccount(
+        IERC6551Registry _registry,
+        address _hatsAccountImplementation,
+        address protocolAddress,
+        uint256 hatId
+    ) internal returns (address) {
+        return
+            _registry.createAccount(
+                _hatsAccountImplementation,
+                SALT,
+                block.chainid,
+                protocolAddress,
+                hatId
+            );
+    }
```
Same as previous comment

dacebt replied:
fixed.

adamgall commented on contracts/DecentHats.sol:
```diff
@@ -0,0 +1,481 @@
+        );
+    }
+
+    function _createElectionEligiblityModule(
+        IHatsModuleFactory hatsModuleFactory,
+        address hatsElectionEligibilityImplementation,
+        uint256 hatId,
+        uint256 topHatId,
+        TermedParam calldata termedParam
+    ) internal returns (address electionModuleAddress) {
+        electionModuleAddress = hatsModuleFactory.createHatsModule(
+            hatsElectionEligibilityImplementation,
+            hatId,
+            abi.encode(topHatId, uint256(0)),
+            abi.encode(termedParam.termEndDateTs),
+            uint256(SALT)
+        );
+    }
```
same as previous comment

dacebt replied:
fixed.

adamgall committed:
nominatedWearer now a single address to match business logic

adamgall committed:
Remove NAME from contract

adamgall committed:
Remove unused import

dacebt committed:
use error and revert

dacebt committed:
remove declaration and pass directly

dacebt committed:
remove _createHat internal func and call directly

dacebt committed:
remove _createElectionEligiblityModule and use function call directly

dacebt committed:
remove _createAccount and use function call directly

dacebt committed:
add comments

dacebt committed:
update to customError for test

dacebt committed:
update name to make sense

adamgall commented on contracts/DecentHats.sol:
```diff
@@ -0,0 +1,436 @@
+     * to avoid a race condition where not more than one active proposal to create a new role can exist at a time.
+     * See: https://github.com/decentdao/decent-interface/issues/2402
+     */
+    function createRoleHat(CreateRoleHatParams calldata params) external {
+        _createHatAndAccountAndMintAndStreams(
+            params.hatsProtocol,
+            params.registry,
+            params.topHatAccount,
+            params.hatsAccountImplementation,
+            params.adminHatId,
+            params.hat
+        );
+
+        params.hatsProtocol.transferHat(
+            params.topHatId,
+            address(this),
+            msg.sender
+        );
+    }
```
I think the last thing to do in this contract, is clean up this standalone function.

@DarksightKellar this one is on you.

So, the reason why we need to transfer the top back back to the safe, is because _this_ contract is the current wearer of the top hat by the time this function is being called. We did that (transaction was created in the frontend and part of the bundle of txs in the proposal that this call is in) because this contract can't create a new hat on a tree unless it's an admin of the hat it's trying to create. Makes sense.

BUT, instead of transferring the top hat to this contract, why not utilize the fact that we've added this contract as a module on the Safe! As of right now, adding this contract as a module on the Safe (in the context of minting a new hat via this call), _is completely unnecessary_, _unless we actually call `execTransactionFromModule` in this function call_.

So let's do that instead of transferring the top hat around.

Please proxy the call to create a new hat through the Safe via `execTransactionFromModule`, and remove the `transferHat` instruction. In the frontend, no need to transfer the top hat to this contract.

adamgall commented on contracts/DecentHats.sol:
```diff
@@ -0,0 +1,436 @@
+     * to avoid a race condition where not more than one active proposal to create a new termed role can exist at a time.
+     * See: https://github.com/decentdao/decent-interface/issues/2402
+     */
+    function createTermedRoleHat(
+        CreateTermedRoleHatParams calldata params
+    ) external {
+        _createTermedHatAndAccountAndMintAndStreams(
+            params.hatsProtocol,
+            params.topHatAccount,
+            params.hatsModuleFactory.createHatsModule( // Create election module and set as eligibility
+                params.hatsElectionEligibilityImplementation,
+                params.hatsProtocol.getNextId(params.adminHatId),
+                abi.encode(params.topHatId, uint256(0)), // [BALLOT_BOX_ID, ADMIN_HAT_ID]
+                abi.encode(params.hat.termedParam.termEndDateTs),
+                uint256(SALT)
+            ),
+            params.adminHatId,
+            params.hat
+        );
+
+        params.hatsProtocol.transferHat(
+            params.topHatId,
+            address(this),
+            msg.sender
+        );
+    }
```
Oh look this function exists too!

@Da-Colon @DarksightKellar see my previous comment in this PR. Same applies here.

dacebt replied:
Let's talk about this today.

adamgall replied:
ready when you are 🫡

adamgall reviewed (changes requested):
Last set of change requests, related to the two external standalone functions which create new Hats on existing trees.

Those functions _shouldn't_ need to transfer the top hat back to the Safe, because this contract _shouldn't_ need to take ownership of the top hat in the first place.

Just use the fact that we add this contract as a module on the Safe, to proxy those calls to create new hats through the Safe.

adamgall committed:
Rename contract to DecentHatsCreationModule

adamgall committed:
Remove code for modifying an existing hats tree (adding a hat)

adamgall committed:
Remove "modifiying tree" (adding hats) external functions from contract, and do some light refactoring

adamgall committed:
Move functions around, add some comments

adamgall committed:
Remove unneeded struct via refactoring

adamgall committed:
Rename "election eligibility" to "elections eligibility", and "registry" to "erc6551Registry"
